### PR TITLE
Add support for NumPy types in Term construction.

### PR DESCRIPTION
**Pull request by ricardo-espinoza**

### CODE REVIEW

File: azure-quantum/azure/quantum/optimization/term.py [hunk 1]
```diff
@@ -4,23 +4,45 @@
 ##
 
 from typing import List, Union, Optional
+import numpy as np
```
Comment (Mobius5150) [inline azure-quantum/azure/quantum/optimization/term.py:7]: Does this add numpy as a dependency?

Comment (ricardo-espinoza) [reply azure-quantum/azure/quantum/optimization/term.py:7]: Yes, after following up with you offline, we're revisiting the approach for this change.

Comment (cgranade) [reply azure-quantum/azure/quantum/optimization/term.py:7]: That can be mitigated by guarding with an `if typing.TYPE_CHECKING` statement: https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

File: azure-quantum/azure/quantum/optimization/term.py [hunk 2]
```diff
@@ -4,23 +4,45 @@
...
 
 __all__ = ['Term']
 
+numpy_integer_types = [np.byte, np.ubyte, np.short, np.ushort, np.intc, np.uintc,
+                       np.int_, np.uint, np.longlong, np.ulonglong, np.int8, np.uint8,
+                       np.int16, np.uint16, np.int32, np.uint32, np.int64, np.uint64]
+
+numpy_float_types = [np.float16, np.float32, np.float64, np.float_, np.half, np.single, np.double]
+
+def _convert_if_numpy_type(param : npt.DTypeLike):
```
Comment (cgranade) [inline azure-quantum/azure/quantum/optimization/term.py:18]: I suspect you want ArrayLike and not DTypeLike here? DTypeLike matches objects representing dtypes, and not values of those dtypes.

File: azure-quantum/azure/quantum/optimization/term.py (hunk 1 above)
Comment (ricardo-espinoza) [reply azure-quantum/azure/quantum/optimization/term.py]: Per discussion offline, we're leaving the numpy dependency in, however, we check for numpy.typing conditionally to support this logic.

File: azure-quantum/azure/quantum/optimization/term.py [hunk 3]
```diff
@@ -4,23 +4,45 @@
...
+    # Attempt first a conversion to a supported type if parameter is a numpy float/int.
+    if (type(param) in numpy_integer_types + numpy_float_types):
```
[the remark this answers is not in the document]
Comment (ricardo-espinoza) [reply azure-quantum/azure/quantum/optimization/term.py]: My understanding is that we would only expect a value in this case, however, changing to npt.ArrayLike to follow the rest of the code.

File: azure-quantum/azure/quantum/optimization/term.py (hunk 2 above)
Comment (ricardo-espinoza) [reply azure-quantum/azure/quantum/optimization/term.py]: Thanks!

File: azure-quantum/azure/quantum/optimization/term.py (hunk 1 above)
Comment (cgranade) [reply azure-quantum/azure/quantum/optimization/term.py]: Fair enough! NumPy is _extremely_ common in scientific Python audiences (may be nearly universal, even), but perhaps less so outside of scientific usage. If we believe that the primary users are likely to be scientific in nature, taking a NumPy dependency is nearly zero-weight as a result.

File: azure-quantum/azure/quantum/optimization/term.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cgranade) [reply azure-quantum/azure/quantum/optimization/term.py]: May be able to resolve by looking at `param.dtype` in the case that `hasattr(param, 'dtype')`.

File: azure-quantum/azure/quantum/optimization/term.py [hunk 4]
```diff
@@ -9,20 +9,52 @@
...
+
+    def _convert_if_numpy_type(param : WArray):
+        # Attempt first a conversion to a supported type if parameter is a numpy float/int.
+        numpy_integer_types = [np.byte, np.ubyte, np.short, np.ushort, np.intc, np.uintc,
+                               np.int_, np.uint, np.longlong, np.ulonglong, np.int8, np.uint8,
+                               np.int16, np.uint16, np.int32, np.uint32, np.int64, np.uint64]
+
+        numpy_float_types = [np.float16, np.float32, np.float64, np.float_, np.half, np.single, np.double]
+
+        if (type(param) in numpy_integer_types + numpy_float_types):
```
Comment (cgranade) [inline azure-quantum/azure/quantum/optimization/term.py:24]:
```suggestion
        if (hasattr(param, 'dtype') and param.dtype in numpy_integer_types + numpy_float_types):
```

Comment (cgranade) [inline azure-quantum/azure/quantum/optimization/term.py:24]: I'd also suggest dropping the parens around the conditional here, as per PEP 8 and/or black style guides.

File: azure-quantum/azure/quantum/optimization/term.py [hunk 5]
```diff
@@ -9,20 +9,52 @@
...
 
+        if(type(w) != None and type(c) != None):
+            raise RuntimeError("Cost has been specified multiple times. Please do not specify 'w' if using 'c'.")
+
         if(w != None):
+            # Legacy support if 'w' is used to specify term instead of the expected 'c'.
+            self.w = _convert_if_numpy_type(w)
             if(type(w) != int and type(w) != float):
-                raise RuntimeError("w must be a float or int value.")
+                raise RuntimeError("w must be a float or int value, or a numpy value that can be converted to those.")
```
Comment (cgranade) [inline azure-quantum/azure/quantum/optimization/term.py:50]:
```suggestion
                raise RuntimeError("w must be a float or int value, or a NumPy value that can be converted to those.")
```

File: azure-quantum/azure/quantum/optimization/term.py (hunk 4 above)
Comment (cgranade) [inline azure-quantum/azure/quantum/optimization/term.py:24]: Third suggestion on this line, but this condition doesn't check if an array is scalar-like (shape `()` or `(1,)`) before calling `.item()`. You may be able to give a more actionable error message by doing so.

Comment (ricardo-espinoza) [reply azure-quantum/azure/quantum/optimization/term.py]: Adding handling.

**Review by cgranade — APPROVED** [review]

**Pull request merged**